File: devbricksx-java-annotations/src/main/java/com/dailystudio/devbricksx/annotations/RoomCompanion.java
package com.dailystudio.devbricksx.annotations;

import androidx.room.ForeignKey;
import androidx.room.Index;

import com.sun.org.apache.xpath.internal.operations.Bool;

import java.lang.annotation.ElementType;
import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;
import java.lang.annotation.Target;

/*
 * RoomCompanion is deprecated by RoomCompanion in package com.dailystudio.devbricksx.annotations.data
 */
@Deprecated
@Retention(RetentionPolicy.RUNTIME)
@Target(ElementType.TYPE)
public @interface RoomCompanion {
    String[] primaryKeys();
    boolean autoGenerate() default false;

    Class<?>[] converters() default {};
    ForeignKey[] foreignKeys() default {};
    Index[] indices() default {};

    Class<?> extension() default Void.class;

    // database name
    String database() default "";

    // database version
    int databaseVersion() default 1;

    // database migrations
    Class<?>[] migrations() default {};

    // dao settings
    int pageSize() default 10;

    // repository settings
    boolean repository() default true;
}

File: devbricksx-annotations/src/main/java/com/dailylstudio/devbricksx/annotations/plus/RoomCompanion.kt
package com.dailylstudio.devbricksx.annotations.plus

import androidx.room.ForeignKey
import androidx.room.Index
import kotlin.reflect.KClass

@Retention(AnnotationRetention.RUNTIME)
@Target(AnnotationTarget.ANNOTATION_CLASS, AnnotationTarget.CLASS)
annotation class RoomCompanion(
    val primaryKeys: Array<String> = [],
    val autoGenerate: Boolean = false,
    val converters: Array<KClass<*>> = [],
    val foreignKeys: Array<ForeignKey> = [],
    val indices: Array<Index> = [],
    val extension: KClass<*> = Void::class,  // database name
    val database: String = "",  // database version
    val databaseVersion: Int = 1,  // database migrations
    val migrations: Array<KClass<*>> = [],  // dao settings
    val pageSize: Int = 10,  // repository settings
    val repository: Boolean = true
)